### PR TITLE
repart: BSD lock block device before rereading partition table

### PR DESCRIPTION
**Pull request by daandemeyer**
Let's take an exclusive lock on the block device before telling
fdisk to reread the partition table so avoid transient errors when
something else (like udev) is operating on the block device.

Fixes https://github.com/systemd/systemd/issues/27236

### CODE REVIEW

File: src/partition/repart.c [hunk 1]
```diff
@@ -4884,17 +4884,24 @@ static int context_write_partition_table(Context *context) {
...
+                if (context->from_scratch) {
+                        if (flock(fd, LOCK_EX) < 0)
+                                return log_error_errno(r, "Failed to lock block device: %m");
+
+                        CLEANUP_BSD_UNLOCK(fd);
+
                         r = fdisk_reread_partition_table(context->fdisk_context);
-                else
+                } else
                         r = fdisk_reread_changes(context->fdisk_context, original_table);
```
Comment (yuwata) [inline src/partition/repart.c:4905]: I am not familiar with the difference, but why the locking is unnecessary for `fdisk_reread_changes()`?

**Review by yuwata — COMMENTED** [review]
In `context_load_partition_table()`, the fd may be already locked... Is this safe??

**Comment by poettering** [comment]
this doesn't look right to me, we already have an exclusive lock on the device, via context_load_partition_table(). If you take another one here this will deadlock

**Comment by poettering** [comment]
(and whenever we open a new loopback block device we create it with LOOP_EX, too)

**Comment by DaanDeMeyer** [comment]
Ah oops I forgot we already take a lock, curious where the error in the issue is coming from then, I guess there's something else that isn't taking a lock?

**Comment by poettering** [comment]
i wonder if for some reason we are not going through the codepath that actually locks things in that test case. dunno.

**Pull request closed without merging**